shenyu-plugin/shenyu-plugin-httpclient/src/main/java/org/apache/shenyu/plugin/httpclient/WebClientPlugin.java

```diff
@@ -100,12 +121,86 @@ public boolean skip(final ServerWebExchange exchange) {
+                                        final long timeout,
+                                        final ShenyuContext shenyuContext,
+                                        final int retryTimes) {
+        Mono<ClientResponse> result = clientResponse;
+        for (int i = 0; i < retryTimes; i++) {
+            result = resend(result, method, exclude, exchange, timeout, shenyuContext);
+        }
+        return result;
+    }
+
+    private Mono<ClientResponse> resend(final Mono<ClientResponse> clientResponse,
+                                        final HttpMethod method,
+                                        final Set<URI> exclude,
+                                        final ServerWebExchange exchange,
+                                        final long timeout,
+                                        final ShenyuContext shenyuContext) {
+        //todo How to add backoff ?
+        return clientResponse.onErrorResume(e -> {
+            final String selectorId = exchange.getAttribute(Constants.DIVIDE_SELECTOR_ID);
+            final List<Upstream> upstreamList = UpstreamCacheManager.getInstance().findUpstreamListBySelectorId(selectorId)
+                    .stream().filter(data -> {
+                        final String trimUri = data.getUrl().trim();
+                        for (URI uri : exclude) {
+                            // exclude already called
+                            if ((uri.getHost() + ":" + uri.getPort()).equals(trimUri)) {
+                                return true;
+                            }
+                        }
+                        return false;
+                    }).collect(Collectors.toList());
+            if (CollectionUtils.isEmpty(upstreamList)) {
+                LOG.error("upstream configuration error： {}", GsonUtils.getInstance().toJson(upstreamList));
+                return Mono.error(new ShenyuException(ShenyuResultEnum.CANNOT_FIND_HEALTHY_UPSTREAM_URL.getMsg()));
+            }
+            final String loadBalance = (String) Optional.ofNullable(exchange.getAttribute(Constants.LOAD_BALANCE)).orElseGet(LoadBalanceEnum.RANDOM::getName);
+            String ip = Objects.requireNonNull(exchange.getRequest().getRemoteAddress()).getAddress().getHostAddress();
+            Upstream upstream = LoadBalancerFactory.selector(upstreamList, loadBalance, ip);
+            if (Objects.isNull(upstream)) {
+                LOG.error("has no upstream");
+                return Mono.error(new ShenyuException(ShenyuResultEnum.CANNOT_FIND_HEALTHY_UPSTREAM_URL.getMsg()));
+            }
+            final URI newUri = buildUri(upstream, exchange, shenyuContext);
+            // in order not to affect the next retry call, this uri needs to be excluded
+            exclude.add(newUri);
+            return handleRequest(method, newUri, exchange, timeout);
+        });
+    }
+
+    private URI buildUri(final Upstream upstream,
```

Comment:
can refactor common method， because in URI plugin？

Reply:
> can refactor common method， because in URI plugin？

Yes, most of in `URIPlugin`, I'll try for it.